pkg/storage/stores/tsdb/head_manager.go

```diff
@@ -272,17 +318,6 @@ func (m *HeadManager) Rotate(t time.Time) error {
 	m.activeHeads = nextHeads
```

Comment:
Let us move all the logic of rotating the head to the ticker loop above to keep it in one place. This approach would also help us avoid the possibility of losing the index when the loop is failing to build the index, but the `Append` keeps rotating the index every 15m without knowing about it.

Here is how I imaging loop would work:
* if `prev` is not nil: `// This is for handling the case where the loop rotated the head but failed to build the index.`
   * Build TSDB from the WAL.
   * If the TSDB got built successfully, rotate the head and let the next tick take care of building the index.
* else:
   * Rotate the head and build the index from `prev`.
   * Set `prev` to nil if we built the index successfully.

This change would mean that `Append` would always keep writing to the active head. It would be the loop taking care of rotating the head and building TSDB out of it. If the loop fails to build the TSDB, it will keep retrying it during each tick, while the `Append` could keep writing to the active head.

Here we should make sure we take the exclusive lock on the mutex for as little as possible, e.g. while changing the references of `m.prev*` and `m.active*` or any other operation which can impact concurrent operations. 

We should also add a metric to track the failures and successes in building the index so we can alert on it.

Does it make sense?

Reply:
This sounds good to me.

I think we can merge once the following are addressed:
> Let us move all the logic of rotating the head to the ticker loop above to keep it in one place. 

> if prev is not nil: `// This is for handling the case where the loop rotated the head but failed to build the index.`